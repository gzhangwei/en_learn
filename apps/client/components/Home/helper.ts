import { ref } from "vue";

import { fetchUserRecentCoursePacks } from "~/api/user-course-progress";
import { type UserRecentCoursePack } from "~/types";

const coursePacks = ref<UserRecentCoursePack[]>([]);

export function useRecentCoursePack() {
  async function fetchCoursePacks() {
    coursePacks.value = await fetchUserRecentCoursePacks();
  }

  return {
    fetchCoursePacks,
    coursePacks,
  };
}
