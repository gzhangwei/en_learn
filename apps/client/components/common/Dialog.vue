<script lang="ts" setup>
import { useModal } from "#imports";

defineProps({
  title: {
    type: String,
    default: "",
  },
  content: {
    type: String,
    default: "",
  },
  showCancel: {
    type: Boolean,
    default: false,
  },
  cancelText: {
    type: String,
    default: "取消",
  },
  showConfirm: {
    type: Boolean,
    default: false,
  },
  confirmText: {
    type: String,
    default: "确认",
  },
});

const modal = useModal();
const emit = defineEmits(["cancel", "confirm"]);

async function onCancel() {
  await modal.close();
  emit("cancel");
}
async function onConfirm() {
  await modal.close();
  emit("confirm");
}
</script>

<template>
  <UModal :ui="{ width: 'w-full sm:max-w-lg' }">
    <div class="flex h-52 flex-col justify-between p-6 text-gray-900 dark:text-white">
      <h2 class="mb-8 text-2xl font-bold">{{ title }}</h2>
      <p class="mb-8 text-base text-gray-700 dark:text-gray-300">
        {{ content }}
      </p>
      <div class="flex w-full justify-end space-x-4">
        <UButton
          v-if="showCancel"
          color="gray"
          class="px-6"
          @click="onCancel"
        >
          {{ cancelText || "取消" }}
        </UButton>
        <UButton
          v-if="showConfirm"
          class="px-6"
          @click="onConfirm"
        >
          {{ confirmText || "确认" }}
        </UButton>
      </div>
    </div>
  </UModal>
</template>
