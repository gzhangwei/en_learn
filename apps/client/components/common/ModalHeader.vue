<template>
  <div class="mb-4">
    <h1 class="text-center text-2xl font-bold text-purple-800 dark:text-white">{{ title }}</h1>
    <div class="absolute right-2 top-2">
      <UButton
        color="gray"
        variant="ghost"
        icon="i-heroicons-x-mark-20-solid"
        @click="$emit('close')"
        tabindex="-1"
        :ui="{ color: { gray: { ghost: 'dark:hover:bg-gray-600' } } }"
      />
    </div>
  </div>
</template>

<script setup lang="ts">
defineProps({
  title: {
    type: String,
    required: true,
  },
});
</script>
