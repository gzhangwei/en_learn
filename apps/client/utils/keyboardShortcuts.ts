import { convertMacKey } from "~/composables/user/shortcutKey";

// 添加全局快捷键
export interface Shortcut {
  key: string;
  ctrlKey: boolean;
  metaKey: boolean;
  command: (keyboardEvent: KeyboardEvent) => void;
}

const shortcuts: Shortcut[] = [];

window.addEventListener("keydown", (e: KeyboardEvent) => {
  const matchingShortcuts = findMatchingShortcut(e);

  matchingShortcuts.forEach((shortcut) => {
    shortcut.command(e);
  });
});

function parseKey(keyString: string) {
  const keys = keyString.toLowerCase().split("+");

  const result = {
    key: keys[keys.length - 1], // 取数组最后一个元素作为 key
    ctrlKey: keys.includes("ctrl"),
    metaKey: keys.includes("command"),
  };

  return result;
}

function findMatchingShortcut(event: KeyboardEvent): Shortcut[] {
  return shortcuts.filter((shortcut) => {
    const preciseMatching =
      shortcut.ctrlKey === event.ctrlKey &&
      shortcut.metaKey === event.metaKey &&
      shortcut.key === convertMacKey(event.key).toLowerCase();

    const anyMatching = shortcut.key === "*";

    return preciseMatching || anyMatching;
  });
}

export function parseShortcutKeys(shortcutKeys: string, separator: string = "+") {
  // 如果只有一个字符的 key，将其转换为大写显示
  return shortcutKeys.split(separator).map((key) => (key.length === 1 ? key.toUpperCase() : key));
}

export function createShortcut(key: string, command: Shortcut["command"]): Shortcut {
  return {
    ...parseKey(key),
    command,
  };
}

export function registerShortcut(key: string, command: Shortcut["command"]) {
  const shortcut = createShortcut(key, command);
  shortcuts.push(shortcut);
  return shortcut;
}

export function cancelShortcut(key: string, command: Shortcut["command"]): void;
export function cancelShortcut(shortcut: Shortcut): void;
export function cancelShortcut(key: string): void;
export function cancelShortcut(keyOrShortcut: string | Shortcut, command?: Shortcut["command"]) {
  function normalizeShortcut(): Shortcut | Partial<Shortcut> {
    if (typeof keyOrShortcut === "object") {
      return keyOrShortcut;
    }
    return command ? createShortcut(keyOrShortcut, command) : parseKey(keyOrShortcut);
  }

  let normalShortcut = normalizeShortcut();

  if (typeof keyOrShortcut === "string" && !command) {
    // 如果只传了 key，删除所有匹配的快捷键
    for (let i = shortcuts.length - 1; i >= 0; i--) {
      if (
        shortcuts[i].key === normalShortcut!.key &&
        shortcuts[i].ctrlKey === normalShortcut!.ctrlKey &&
        shortcuts[i].metaKey === normalShortcut!.metaKey
      ) {
        shortcuts.splice(i, 1);
      }
    }
    return;
  }

  const index = shortcuts.findIndex(({ key, command, ctrlKey, metaKey }) => {
    // 精准匹配对应快捷键对象
    return (
      key === normalShortcut.key &&
      ctrlKey === normalShortcut.ctrlKey &&
      metaKey === normalShortcut.metaKey &&
      command === normalShortcut.command
    );
  });

  if (index !== -1) {
    // 删除
    shortcuts.splice(index, 1);
  }
}

export function cleanAllShortcut() {
  shortcuts.length = 0;
}
