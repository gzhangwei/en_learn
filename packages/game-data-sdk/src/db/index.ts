import type { PostgresJsDatabase } from "drizzle-orm/postgres-js";

import { sql } from "drizzle-orm";
import { drizzle } from "drizzle-orm/postgres-js";
import postgres from "postgres";

import type { SchemaType } from "@earthworm/schema";
import { schemas } from "@earthworm/schema";

export type DbType = PostgresJsDatabase<SchemaType>;

// eslint-disable-next-line import/no-mutable-exports
export let db: DbType;
let connection: postgres.Sql;

export const setupDB = async (databaseURL: string) => {
  connection = postgres(databaseURL || "");

  db = drizzle(connection, {
    schema: schemas,
  });

  return db;
};

export async function cleanDB(db: DbType) {
  await db.execute(
    sql`TRUNCATE TABLE courses, statements, "course_packs" , "user_course_progress", "course_history", "user_learn_record", "memberships" RESTART IDENTITY CASCADE;`,
  );
}

export async function teardownDb() {
  if (connection) {
    await connection.end();
  }
}
