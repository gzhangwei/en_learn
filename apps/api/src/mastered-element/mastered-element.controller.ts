import { Body, Controller, Delete, Get, Param, Post, UseGuards } from "@nestjs/common";

import { AuthGuard } from "../guards/auth.guard";
import { User, UserEntity } from "../user/user.decorators";
import { MasteredElementService } from "./mastered-element.service";
import { AddMasteredElementDto } from "./model/add-mastered-element.dto";
import { RemoveMasteredElementDto } from "./model/remove-mastered-element.dto";

@Controller("mastered-elements")
export class MasteredElementController {
  constructor(private readonly masteredElementService: MasteredElementService) {}

  @UseGuards(AuthGuard)
  @Post()
  async addMasteredElement(
    @User() user: UserEntity,
    @Body() addMasteredElementDto: AddMasteredElementDto,
  ) {
    return await this.masteredElementService.addMasteredElement(
      user.userId,
      addMasteredElementDto.content,
    );
  }

  @UseGuards(AuthGuard)
  @Get()
  async getMasteredElements(@User() user: UserEntity) {
    return this.masteredElementService.getMasteredElements(user.userId);
  }

  @UseGuards(AuthGuard)
  @Delete(":elementId")
  async removeMasteredElement(
    @User() user: UserEntity,
    @Param() removeMasteredElementDto: RemoveMasteredElementDto,
  ) {
    return await this.masteredElementService.removeMasteredElement(
      user.userId,
      removeMasteredElementDto.elementId,
    );
  }
}
