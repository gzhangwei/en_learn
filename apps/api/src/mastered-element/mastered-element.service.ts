import { BadRequestException, Inject, Injectable, NotFoundException } from "@nestjs/common";
import { and, desc, eq } from "drizzle-orm";

import { masteredElements as masteredElementsSchema } from "@earthworm/schema";
import { DB, DbType } from "../global/providers/db.provider";

interface ElementContent {
  english: string;
}

@Injectable()
export class MasteredElementService {
  constructor(@Inject(DB) private db: DbType) {}

  async addMasteredElement(userId: string, content: ElementContent) {
    if (!content.english) {
      throw new BadRequestException("Element english content is required");
    }

    if (await this.isMastered(userId, content)) {
      throw new BadRequestException("这个内容已经掌握了");
    }

    const [entity] = await this.db
      .insert(masteredElementsSchema)
      .values({
        userId,
        content: JSON.stringify(content),
        masteredAt: new Date(),
      })
      .returning();

    entity.content = JSON.parse(entity.content as string);
    return entity;
  }

  async getMasteredElements(userId: string) {
    const result = await this.db
      .select()
      .from(masteredElementsSchema)
      .where(eq(masteredElementsSchema.userId, userId))
      .orderBy(desc(masteredElementsSchema.masteredAt));

    return result.map((item) => ({
      ...item,
      content: JSON.parse(item.content as string),
    }));
  }

  async removeMasteredElement(userId: string, elementId: string) {
    const result = await this.db
      .delete(masteredElementsSchema)
      .where(
        and(eq(masteredElementsSchema.userId, userId), eq(masteredElementsSchema.id, elementId)),
      )
      .returning();

    if (result.length === 0) {
      throw new NotFoundException(
        `Mastered element with id ${elementId} not found for user ${userId}`,
      );
    }

    return result[0];
  }

  async isMastered(userId: string, content: ElementContent) {
    const result = await this.db
      .select()
      .from(masteredElementsSchema)
      .where(
        and(
          eq(masteredElementsSchema.userId, userId),
          eq(masteredElementsSchema.content, JSON.stringify(content)),
        ),
      )
      .limit(1);

    return result.length > 0;
  }
}
