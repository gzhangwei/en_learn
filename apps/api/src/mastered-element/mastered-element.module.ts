import { Module } from "@nestjs/common";

import { MasteredElementController } from "./mastered-element.controller";
import { MasteredElementService } from "./mastered-element.service";

@Module({
  providers: [MasteredElementService],
  controllers: [MasteredElementController],
})
export class MasteredElementModule {}
