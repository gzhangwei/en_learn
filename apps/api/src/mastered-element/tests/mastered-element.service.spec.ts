import { BadRequestException, NotFoundException } from "@nestjs/common";
import { Test, TestingModule } from "@nestjs/testing";

import { cleanDB, testImportModules } from "../../../test/helper/utils";
import { endDB } from "../../common/db";
import { DB, DbType } from "../../global/providers/db.provider";
import { MasteredElementService } from "../mastered-element.service";

describe("MasteredElementService", () => {
  let db: DbType;
  let masteredElementService: MasteredElementService;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: testImportModules,
      providers: [MasteredElementService],
    }).compile();

    db = module.get<DbType>(DB);
    masteredElementService = module.get<MasteredElementService>(MasteredElementService);
  });

  afterAll(async () => {
    await cleanDB(db);
    await endDB();
  });

  beforeEach(async () => {
    await cleanDB(db);
  });

  describe("addMasteredElement", () => {
    it("should successfully add a new mastered element", async () => {
      const userId = "testUser";
      const content = { english: "test" };
      const result = await masteredElementService.addMasteredElement(userId, content);
      expect(result).toBeDefined();
      expect(result.content).toEqual(content);
    });

    it("should throw BadRequestException when adding an existing element", async () => {
      const userId = "testUser";
      const content = { english: "test" };
      await masteredElementService.addMasteredElement(userId, content);
      await expect(masteredElementService.addMasteredElement(userId, content)).rejects.toThrow(
        BadRequestException,
      );
    });

    it("should throw BadRequestException when adding element without english property", async () => {
      const userId = "testUser";
      const content = { someOtherProperty: "test" };
      await expect(
        masteredElementService.addMasteredElement(userId, content as any),
      ).rejects.toThrow(BadRequestException);
    });

    it("should correctly handle additional properties in content", async () => {
      const userId = "testUser";
      const content = { english: "test", chinese: "测试" };
      const result = await masteredElementService.addMasteredElement(userId, content);
      expect(result.content).toEqual(content);
    });
  });

  describe("getMasteredElements", () => {
    it("should return all mastered elements for a user", async () => {
      const userId = "testUser";
      const contents = [{ english: "test1" }, { english: "test2" }, { english: "test3" }];
      for (const content of contents) {
        await masteredElementService.addMasteredElement(userId, content);
      }

      const result = await masteredElementService.getMasteredElements(userId);

      expect(result).toHaveLength(3);
      expect(result.map((r) => r.content)).toMatchSnapshot();
    });

    it("should return elements in descending order of masteredAt", async () => {
      const userId = "testUser";
      const contents = [{ english: "test1" }, { english: "test2" }, { english: "test3" }];
      for (const content of contents) {
        await masteredElementService.addMasteredElement(userId, content);
      }
      const result = await masteredElementService.getMasteredElements(userId);
      expect(result.map((r) => r.content)).toMatchSnapshot();
    });

    it("should return an empty array when user has no mastered elements", async () => {
      const result = await masteredElementService.getMasteredElements("nonexistentUser");
      expect(result).toEqual([]);
    });
  });

  describe("removeMasteredElement", () => {
    it("should successfully remove an existing mastered element", async () => {
      const userId = "testUser";
      const content = { english: "test" };
      const addedElement = await masteredElementService.addMasteredElement(userId, content);
      await masteredElementService.removeMasteredElement(userId, addedElement.id);
      const elements = await masteredElementService.getMasteredElements(userId);
      expect(elements).toHaveLength(0);
    });

    it("should throw NotFoundException when removing a non-existent element", async () => {
      const userId = "testUser";
      const nonExistentId = "nonexistent";
      await expect(
        masteredElementService.removeMasteredElement(userId, nonExistentId),
      ).rejects.toThrow(NotFoundException);
    });
  });

  describe("Error handling and edge cases", () => {
    it("should handle removal with invalid elementId", async () => {
      const userId = "testUser";
      const invalidElementId = "";
      await expect(
        masteredElementService.removeMasteredElement(userId, invalidElementId),
      ).rejects.toThrow();
    });
  });
});
