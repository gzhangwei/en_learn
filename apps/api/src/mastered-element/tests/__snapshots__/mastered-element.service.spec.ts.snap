// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MasteredElementService getMasteredElements should return all mastered elements for a user 1`] = `
[
  {
    "english": "test3",
  },
  {
    "english": "test2",
  },
  {
    "english": "test1",
  },
]
`;

exports[`MasteredElementService getMasteredElements should return elements in descending order of masteredAt 1`] = `
[
  {
    "english": "test3",
  },
  {
    "english": "test2",
  },
  {
    "english": "test1",
  },
]
`;
